File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
import 'dart:collection';
class Solution {
  TreeNode? createBinaryTree(List<List<int>> descriptions) {
    final children = <int>{};
    final edges = <int, List<(int, int)>>{};
    for (final desc in descriptions){
      children.add(desc[1]);
      edges[desc[0]] ??= [];
      edges[desc[0]]!.add((desc[1], desc[2]));
    }
    final rootVal = edges.keys.firstWhere((k) => !children.contains(k));
    final root = TreeNode(rootVal);
    final q = Queue<TreeNode>()..add(root);
    while (q.isNotEmpty){
      final cur = q.removeFirst();
      final next = edges[cur.val] ?? [];
      for (final (n, left) in next){
        if (left == 1){
          cur.left = TreeNode(n);
          q.add(cur.left!);
        } else {
          cur.right = TreeNode(n);
          q.add(cur.right!);
        }
      }
    }
    return root;
  }
}

File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.ex
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @spec create_binary_tree(descriptions :: [[integer]]) :: TreeNode.t | nil
  def create_binary_tree(descriptions) do
    {children, edges} = descriptions
    |> Enum.reduce({MapSet.new(), %{}}, fn [p, c, left], {children, edges} ->
      edge = Map.get(edges, p, %{})
      {MapSet.put(children, c), Map.put(edges, p, Map.put(edge, left, c))}
    end)
    
    root = edges
    |> Map.drop(children)
    |> Map.keys()
    |> hd()

    helper(edges, root)
  end
  defp helper(edges, root) do
    case Map.get(edges, root) do
      %{1 => left, 0 => right} -> %TreeNode{val: root, left: helper(edges, left), right: helper(edges, right)}
      %{1 => left} -> %TreeNode{val: root, left: helper(edges, left)}
      %{0 => right} -> %TreeNode{val: root, right: helper(edges, right)}
      nil -> %TreeNode{val: root}
    end
  end
end